react-renderer/src/Evaluator.ts

```diff
@@ -7,11 +7,12 @@ import {
   mapKeys,
   concat,
   PropertyPath,
+  zip,
 } from "lodash";
 import { mapValues } from "lodash";
 import { dist, randFloat } from "./Util";
 import seedrandom from "seedrandom";
-import { Tensor, Variable, scalar, pad2d } from "@tensorflow/tfjs-node";
+import { Tensor, Variable, scalar, pad2d } from "@tensorflow/tfjs";
```

Comment:
Is there a clear difference between the tfjs node version vs pure js version? 

Regarding backends, if possible we should use the [react native](https://github.com/tensorflow/tfjs/tree/master/tfjs-react-native) version of tfjs, since it supports model inference and training via GPU/WebGL, which is the fastest option. Does our current frontend work with react native?

There is also an alpha wasm backend [here](https://github.com/tensorflow/tfjs/tree/master/tfjs-backend-wasm) that only supports inference (so we can't use it, since we want "training").

In terms of speed, the wasm page has useful benchmarks for specific models. Basically plain JS is the slowest. Quoting:

* "For this model, the WASM backend is between ~3X-11.5X faster than the plain JS backend, and ~5.3-7.7X slower than the WebGL backend."
* "You should always try to use the WASM backend over the plain JS backend since it is strictly faster on all devices, across all model sizes. Compared to the WebGL backend, the WASM backend has better numerical stability, and wider device support. Performance-wise, our benchmarks show that:
  * "For medium-sized models (~100-500M multiply-adds), the WASM backend is several times slower than the WebGL backend."
  * "For lite models (~20-60M multiply-adds), the WASM backend has comparable performance to the WebGL backend"

Reply:
@hypotext React native is a runtime for react that runs on mobile devices, so I assume tfjs-rn is based on native code and hooks into mobile-specific chips. I would expect hardware acceleration to still be available with the tfjs we use now, check this page: https://www.tensorflow.org/js/guide/platform_environment#backends

Reply:
That's a better source, thanks @maxkrieger. Yes, let's just make sure we use the WebGL backend, and do the memory management that that requires (I'm happy to work on this)

Reply:
I’m aware of these backend options and saw some resource management tricks. You should feel free to try them out! My purpose on this branch is to get the optimizer working with the evaluator, so performance wasn’t my main concern